---
title: "mirna-quant-predict"
date: "2026-02-09"
---

本流程說明如何從 small RNA-seq FASTQ 檔案進行 miRNA 定量（quantification），並進一步執行 miRNA target prediction（RNAhybrid 與 miRanda）。

# miRNA 定量

::: {.callout-important}
small RNA-seq 前處理與 miRNA 定量部分，已整理為完整 workflow 並公開於 [https://github.com/benson1231/mirdeep2](https://github.com/benson1231/mirdeep2)（基於 miRDeep2 工具進行調整與最佳化，適用於人類檢體分析）。詳細流程、腳本與範例資料請參考該 repository。
:::

# miRNA 標的預測工具

::: {.callout-note}
miRNA target prediction 在研究上大致可分為三種主要方向：

* 序列互補性與結合能量預測演算法，例如：RNAhybrid、miRanda
* 演化保守性或機器學習模型預測，例如：TargetScan、miRDB
* 實驗驗證資料庫，例如：miRTarBase、starBase

實務上通常會與資料庫或其他預測工具交叉比對，以降低假陽性。
:::

本步驟說明如何利用 `RNAhybrid` 與 `miRanda` 演算法進行 miRNA–mRNA 配對分析，以預測潛在 target genes。

## 創建 conda 環境

```bash
mamba env create -f https://raw.githubusercontent.com/benson1231/tools/main/envs/mirna-predict.yaml
mamba activate mirna-predict

mkdir results/
```

## 使用 `RNAhybrid` 演算法

::: {.callout-tip}
## Reference

Krüger, Jan, and Marc Rehmsmeier. “RNAhybrid: microRNA target prediction easy, fast and flexible.” Nucleic acids research vol. 34,Web Server issue (2006): W451-4. doi:10.1093/nar/gkl243
:::

- [RNAhybrid 使用手冊](https://manpages.ubuntu.com/manpages/jammy/man1/RNAhybrid.1.html) 

```bash
RNAhybrid -s 3utr_human -t data/target.fa -q data/mirna.fa > results/hybrid.out
```

::: {.callout-note collapse="true"}
## RNAhybrid -h（點擊展開結果）

```
Usage: RNAhybrid [options] [target sequence] [query sequence].

options:

  -b <number of hits per target>
  -c compact output
  -d <xi>,<theta>
  -f helix constraint
  -h help
  -m <max targetlength>
  -n <max query length>
  -u <max internal loop size (per side)>
  -v <max bulge loop size>
  -e <energy cut-off>
  -p <p-value cut-off>
  -s (3utr_fly|3utr_worm|3utr_human)
  -g (ps|png|jpg|all)
  -t <target file>
  -q <query file>

Either a target file has to be given (FASTA format)
or one target sequence directly.

Either a query file has to be given (FASTA format)
or one query sequence directly.

The helix constraint format is "from,to", eg. -f 2,7 forces
structures to have a helix from position 2 to 7 with respect to the query.

<xi> and <theta> are the position and shape parameters, respectively,
of the extreme value distribution assumed for p-value calculation.
If omitted, they are estimated from the maximal duplex energy of the query.
In that case, a data set name has to be given with the -s flag.


PS graphical output not supported.
```
:::

## 使用 `miranda` 演算法

::: {.callout-tip}
## Reference

Enright, Anton J et al. “MicroRNA targets in Drosophila.” Genome biology vol. 5,1 (2003): R1. doi:10.1186/gb-2003-5-1-r1
:::

```bash
miranda data/mirna.fa data/target.fa -out results/miranda.out
```

::: {.callout-note collapse="true"}
## miranda --help（點擊展開結果）

```
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
miranda v3.3a    microRNA Target Scanning Algorithm
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
(c) 2003 Memorial Sloan-Kettering Cancer Center, New York

Authors: Anton Enright, Bino John, Chris Sander and Debora Marks
(mirnatargets (at) cbio.mskcc.org - reaches all authors)

Software written by: Anton Enright
Distributed for anyone to use under the GNU Public License (GPL),
See the files 'COPYING' and 'LICENSE' for details

If you use this software please cite:
Enright AJ, John B, Gaul U, Tuschl T, Sander C and Marks DS;
(2003) Genome Biology; 5(1):R1.

   miranda comes with ABSOLUTELY NO WARRANTY;
   This is free software, and you are welcome to redistribute it
   under certain conditions; type `miranda --license' for details.

miRanda is an miRNA target scanner which aims to predict mRNA
targets for microRNAs using dynamic-programming alignment and
thermodynamics.

Usage:  miranda file1 file2 [options..]

Where:
        'file1' is a FASTA file with a microRNA query
        'file2' is a FASTA file containing the sequence(s)
        to be scanned.

OPTIONS

 --help -h      Display this message
 --version -v   Display version information
 --license      Display license information

Core algorithm parameters:
 -sc S          Set score threshold to S                [DEFAULT: 140.0]
 -en -E         Set energy threshold to -E kcal/mol     [DEFAULT: 1.0]
 -scale Z       Set scaling parameter to Z              [DEFAULT: 4.0]
 -strict        Demand strict 5' seed pairing           [DEFAULT: off]

Alignment parameters:
 -go -X         Set gap-open penalty to -X              [DEFAULT: -4.0]
 -ge -Y         Set gap-extend penalty to -Y            [DEFAULT: -9.0]

General Options:
 -out file      Output results to file                  [DEFAULT: off]
 -quiet         Output fewer event notifications        [DEFAULT: off]
 -trim T        Trim reference sequences to T nt        [DEFAULT: off]
 -noenergy      Do not perform thermodynamics           [DEFAULT: off]

 -restrict file Restricts scans to those between
                specific miRNAs and UTRs
                provided in a pairwise
                tab-separated file                      [DEFAULT: off]

Other Options:
 -keyval        Key value pairs ??                      [DEFAULT:]


This software will be further developed under the open source model,
coordinated by Anton Enright and Chris Sander (miranda (at) cbio.mskcc.org).

Please send bug reports to: miranda (at) cbio.mskcc.org.
```
:::

---

# miRNA 標的預測平台

## TargetScan

[TargetScan](https://www.targetscan.org/vert_80/)

## miRWalk

[miRWalk](http://mirwalk.umm.uni-heidelberg.de/)

## miRDB

[miRDB]([miRDB](https://mirdb.org/))

---

# miRNA 資料庫

## miRTarBase

[miRTarBase](https://awi.cuhk.edu.cn/~miRTarBase/miRTarBase_2025/php/index.php)

::: {.callout-tip}
## Reference

Cui, Shidong et al. “miRTarBase 2025: updates to the collection of experimentally validated microRNA-target interactions.” Nucleic acids research vol. 53,D1 (2025): D147-D156. doi:10.1093/nar/gkae1072
:::

---

# 綜合整理

## Tools4miRs

[Tools4miRs](https://tools4mirs.org/) 是一個很詳盡的 miRNA 分析方法整理平台，彙總了所有處理 miRNA 可能會用到工具，並詳細比較各個工具的特性，非常值得參考。

::: {.callout-tip}
## Reference

[Lukasik A, Wójcikowski M, Zielenkiewicz P.
Tools4miRs – one place to gather all the tools for miRNA analysis. Bioinformatics. 2016;
doi:10.1093/bioinformatics/btw189](https://academic.oup.com/bioinformatics/article/32/17/2722/2450725)
:::