---
title: "RNA-seq deseq2 workflow"
date: "2026-01-24"
---

::: callout-tip
## Reference

Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15:550. 10.1186/s13059-014-0550-8

本篇參考 [bioconductor's workflow](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
:::

::: {.callout-important}
請使用 un-normalized counts 進行分析!
DESeq2 模型內部會對 library size 進行校正，因此不應使用經過 transformed 或 normalized 的值作為輸入!
:::

# 使用 tximport 匯入(方案一)

首先需要樣本的 metadata，紀錄樣本的資訊包含 `sample name`, `batch`, `condition` 以及其他重要資訊

```{r}
library("tximport")
library("readr")
library("tximportData")
dir <- system.file("extdata", package="tximportData")
samples <- read.table(file.path(dir,"samples.txt"), header=TRUE)
samples$condition <- factor(rep(c("A","B"),each=3))
rownames(samples) <- samples$run
samples[,c("pop","center","run","condition")]
```

```
          pop center       run condition
ERR188297 TSI  UNIGE ERR188297         A
ERR188088 TSI  UNIGE ERR188088         A
ERR188329 TSI  UNIGE ERR188329         A
ERR188288 TSI  UNIGE ERR188288         B
ERR188021 TSI  UNIGE ERR188021         B
ERR188356 TSI  UNIGE ERR188356         B
```

接著透過以下方式列出檔案位置，並創建 named-list，key 是樣本名稱，value 是檔案位置

```{r}
files <- file.path(dir,"salmon", samples$run, "quant.sf.gz")
names(files) <- samples$run
files
```

```
                                                                                              ERR188297 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188297/quant.sf.gz" 
                                                                                              ERR188088 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188088/quant.sf.gz" 
                                                                                              ERR188329 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188329/quant.sf.gz" 
                                                                                              ERR188288 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188288/quant.sf.gz" 
                                                                                              ERR188021 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188021/quant.sf.gz" 
                                                                                              ERR188356 
"C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188356/quant.sf.gz" 
```

## 準備 tx2gene 對照表

另外由於原始資訊是 `ENST` transcript 為編號，後續若要合併到基因層級需要對照表 `tx2gene`。

```{r}
tx2gene <- read_csv(file.path(dir, "tx2gene.gencode.v27.csv"))
head(tx2gene)
```

```
  TXNAME            GENEID           
  <chr>             <chr>            
1 ENST00000456328.2 ENSG00000223972.5
2 ENST00000450305.2 ENSG00000223972.5
3 ENST00000473358.1 ENSG00000243485.5
4 ENST00000469289.1 ENSG00000243485.5
5 ENST00000607096.1 ENSG00000284332.1
6 ENST00000606857.1 ENSG00000268020.3
```

## 匯入 count 資料

將 `files` 的檔案匯入，給予 `tx2gene` 會將 `ENST` 的 script-id 轉成 gene level的 `ENSG`，後續用基因層級分析 

```{r}
txi <- tximport(files, type="salmon", tx2gene=tx2gene)
```

```
reading in files with read_tsv
1 2 3 4 5 6 
summarizing abundance
summarizing counts
summarizing length
```

## 建立 DESeq2 物件

建立 `DESeqDataSet` 物件，

```{r}
library("DESeq2")
dds_1 <- DESeqDataSetFromTximport(txi,
                                  colData = samples,
                                  design = ~ condition)
```

```
using counts and average transcript lengths from tximport
```

---

# 使用 tximeta 匯入(方案二)

::: {.callout-note}
`Tximeta` 是 `tximport` 的功能擴充版，可以自動添加 metadata
:::

## 創建 coldata

```{r}
coldata <- samples
coldata$files <- files
coldata$names <- coldata$run

head(coldata)
```

```
          pop center                assay    sample experiment       run condition
ERR188297 TSI  UNIGE NA20503.1.M_111124_5 ERS185497  ERX163094 ERR188297         A
ERR188088 TSI  UNIGE NA20504.1.M_111124_7 ERS185242  ERX162972 ERR188088         A
ERR188329 TSI  UNIGE NA20505.1.M_111124_6 ERS185048  ERX163009 ERR188329         A
ERR188288 TSI  UNIGE NA20507.1.M_111124_7 ERS185412  ERX163158 ERR188288         B
ERR188021 TSI  UNIGE NA20508.1.M_111124_2 ERS185362  ERX163159 ERR188021         B
ERR188356 TSI  UNIGE NA20514.1.M_111124_4 ERS185217  ERX163062 ERR188356         B
                                                                                                          files     names
ERR188297 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188297/quant.sf.gz ERR188297
ERR188088 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188088/quant.sf.gz ERR188088
ERR188329 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188329/quant.sf.gz ERR188329
ERR188288 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188288/quant.sf.gz ERR188288
ERR188021 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188021/quant.sf.gz ERR188021
ERR188356 C:/Users/benson_lee/AppData/Local/R/win-library/4.4/tximportData/extdata/salmon/ERR188356/quant.sf.gz ERR188356
```

## 讀檔匯入

此步驟不需提供 `tx2gene`，因為 `tximeta` 本身會自動添加註釋

```{r}
library("tximeta")
se <- tximeta(coldata)
```

```
importing quantifications
reading in files with read_tsv
1 2 3 4 5 6 
found matching transcriptome:
[ GENCODE - Homo sapiens - release 27 ]
loading existing TxDb created: 2026-01-23 07:59:39
loading existing transcript ranges created: 2026-01-23 07:59:40
```

```{r}
se
```

```
class: RangedSummarizedExperiment 
dim: 200401 6 
metadata(6): tximetaInfo quantInfo ... txomeInfo txdbInfo
assays(3): counts abundance length
rownames(200401): ENST00000456328.2 ENST00000450305.2 ... ENST00000387460.2 ENST00000387461.2
rowData names(3): tx_id gene_id tx_name
colnames(6): ERR188297 ERR188088 ... ERR188021 ERR188356
colData names(8): pop center ... condition names
```

## 建立 DESeqDataSet 物件

```{r}
dds_2 <- DESeqDataSet(se, design = ~ condition)
```

```
using counts and average transcript lengths from tximeta
```

---

# 直接匯入count matrix(方案三)

```{r}
library(DESeq2)

cts <- read.csv(
  "data/airway_counts.csv",
  row.names = 1,
  check.names = FALSE
)

cts <- as.matrix(cts)
head(cts)
```

```
                SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516 SRR1039517 SRR1039520 SRR1039521
ENSG00000000003        679        448        873        408       1138       1047        770        572
ENSG00000000005          0          0          0          0          0          0          0          0
ENSG00000000419        467        515        621        365        587        799        417        508
ENSG00000000457        260        211        263        164        245        331        233        229
ENSG00000000460         60         55         40         35         78         63         76         60
ENSG00000000938          0          0          2          0          1          0          0          0
```

```{r}
coldata <- read.csv(
  "data/airway_coldata.csv",
  row.names = 1
)

coldata$dex  <- factor(coldata$dex)
coldata$cell <- factor(coldata$cell)
coldata
```

```
           SampleName    cell   dex albut        Run avgLength Experiment    Sample    BioSample
SRR1039508 GSM1275862  N61311 untrt untrt SRR1039508       126  SRX384345 SRS508568 SAMN02422669
SRR1039509 GSM1275863  N61311   trt untrt SRR1039509       126  SRX384346 SRS508567 SAMN02422675
SRR1039512 GSM1275866 N052611 untrt untrt SRR1039512       126  SRX384349 SRS508571 SAMN02422678
SRR1039513 GSM1275867 N052611   trt untrt SRR1039513        87  SRX384350 SRS508572 SAMN02422670
SRR1039516 GSM1275870 N080611 untrt untrt SRR1039516       120  SRX384353 SRS508575 SAMN02422682
SRR1039517 GSM1275871 N080611   trt untrt SRR1039517       126  SRX384354 SRS508576 SAMN02422673
SRR1039520 GSM1275874 N061011 untrt untrt SRR1039520       101  SRX384357 SRS508579 SAMN02422683
SRR1039521 GSM1275875 N061011   trt untrt SRR1039521        98  SRX384358 SRS508580 SAMN02422677
```

::: {.callout-important}
此處要注意count matrix 與 colData 的順序需一致!
:::

```{r}
dds_3 <- DESeqDataSetFromMatrix(
  countData = cts,
  colData   = coldata,
  design    = ~ cell + dex
)
dds_3
```

```
class: DESeqDataSet 
dim: 63677 8 
metadata(1): version
assays(1): counts
rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492 ENSG00000273493
rowData names(0):
colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
colData names(9): SampleName cell ... Sample BioSample
```

---

# 直接使用 SummarizedExperiment 物件(方案四)

若已經有 RangedSummarizedExperiment 物件則可直接導入 DESeqDataSet 物件

```{r}
library("airway")
data("airway")
se <- airway
se
```

```
class: RangedSummarizedExperiment 
dim: 63677 8 
metadata(1): ''
assays(1): counts
rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492 ENSG00000273493
rowData names(10): gene_id gene_name ... seq_coord_system symbol
colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
colData names(9): SampleName cell ... Sample BioSample
```

```{r}
library("DESeq2")
dds <- DESeqDataSet(se, design = ~ cell + dex)
dds
```

```
class: DESeqDataSet 
dim: 63677 8 
metadata(2): '' version
assays(1): counts
rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492 ENSG00000273493
rowData names(10): gene_id gene_name ... seq_coord_system symbol
colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
colData names(9): SampleName cell ... Sample BioSample
```

---

# 預先過濾

繼續以方案四的 `dds` 物件分析

```{r}
smallestGroupSize <- 3
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
```

---

# 設定 level

需指定以誰當基準，不然 R 預設是字母順序

```{r}
colData(dds)
```

```
DataFrame with 8 rows and 9 columns
           SampleName     cell      dex    albut        Run avgLength Experiment    Sample    BioSample
             <factor> <factor> <factor> <factor>   <factor> <integer>   <factor>  <factor>     <factor>
SRR1039508 GSM1275862  N61311     untrt    untrt SRR1039508       126  SRX384345 SRS508568 SAMN02422669
SRR1039509 GSM1275863  N61311     trt      untrt SRR1039509       126  SRX384346 SRS508567 SAMN02422675
SRR1039512 GSM1275866  N052611    untrt    untrt SRR1039512       126  SRX384349 SRS508571 SAMN02422678
SRR1039513 GSM1275867  N052611    trt      untrt SRR1039513        87  SRX384350 SRS508572 SAMN02422670
SRR1039516 GSM1275870  N080611    untrt    untrt SRR1039516       120  SRX384353 SRS508575 SAMN02422682
SRR1039517 GSM1275871  N080611    trt      untrt SRR1039517       126  SRX384354 SRS508576 SAMN02422673
SRR1039520 GSM1275874  N061011    untrt    untrt SRR1039520       101  SRX384357 SRS508579 SAMN02422683
SRR1039521 GSM1275875  N061011    trt      untrt SRR1039521        98  SRX384358 SRS508580 SAMN02422677
```

有兩種方法指定 Reference，一是直接對該欄設定 factor level，二是使用 relevel 設定

```{r}
# 方法一
dds$dex <- factor(dds$dex, levels = c("untrt","trt"))

# 方法二
dds$dex <- relevel(dds$dex, ref = "untrt")
```

::: {.callout-tip}
也可使用 droplevels 移除 level，但比較少用

dds$condition <- droplevels(dds$condition)
:::

# 差異表達分析

Differential expression analysis

```{r}
dds <- DESeq(dds)
```

```
using pre-existing size factors
estimating dispersions
found already estimated dispersions, replacing these
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
```

使用 `results` 提取結果

```{r}
res <- results(dds)
res
```

```
log2 fold change (MLE): dex trt vs untrt 
Wald test p-value: dex trt vs untrt 
DataFrame with 16596 rows and 6 columns
                 baseMean log2FoldChange     lfcSE      stat      pvalue        padj
                <numeric>      <numeric> <numeric> <numeric>   <numeric>   <numeric>
ENSG00000000003   709.880     -0.3839261 0.1008515 -3.806846 1.40750e-04 1.09455e-03
ENSG00000000419   521.156      0.2041705 0.1116546  1.828590 6.74611e-02 1.85852e-01
ENSG00000000457   237.573      0.0352858 0.1412422  0.249825 8.02723e-01 9.04035e-01
ENSG00000000460    58.035     -0.0923157 0.2792981 -0.330527 7.41002e-01 8.71680e-01
ENSG00000000971  5826.538      0.4237405 0.0893546  4.742236 2.11372e-06 2.48262e-05
...                   ...            ...       ...       ...         ...         ...
ENSG00000273448  14.02944      0.0583225  0.483593  0.120603    0.904006    0.954503
ENSG00000273472  11.08483     -0.4087635  0.558196 -0.732294    0.463989    0.675274
ENSG00000273486  15.47814     -0.1516640  0.471121 -0.321922    0.747512    0.874323
ENSG00000273487   8.17605      1.0408751  0.678127  1.534926    0.124802    0.292586
ENSG00000273488   8.59778      0.1088526  0.618005  0.176136    0.860187    0.933030
```

可以透過 `resultsNames` 得到有哪些比較

```{r}
resultsNames(dds)
```

```
[1] "Intercept"               "cell_N061011_vs_N052611" "cell_N080611_vs_N052611" "cell_N61311_vs_N052611"  "dex_trt_vs_untrt"  
```

針對不同條件比較，可以使用 `name` 或 `contrast` 指定要輸出哪個比較結果，以下兩個指令是等效的

::: {.callout-important}
contrast 為三個字串的 `list`，最前面是變項名稱，接著是實驗組名稱，最後是對照組(Reference)名稱
:::

```{r}
res <- results(dds, name="dex_trt_vs_untrt")
res <- results(dds, contrast=c("dex","trt","untrt"))
```

# Log fold change shrinkage(可選)

```{r}
resLFC <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm")
resLFC
```

```
using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
log2 fold change (MAP): dex trt vs untrt 
Wald test p-value: dex trt vs untrt 
DataFrame with 16596 rows and 5 columns
                 baseMean log2FoldChange     lfcSE      pvalue        padj
                <numeric>      <numeric> <numeric>   <numeric>   <numeric>
ENSG00000000003   709.880     -0.3672194 0.1002141 1.40750e-04 1.09455e-03
ENSG00000000419   521.156      0.1843404 0.1077340 6.74611e-02 1.85852e-01
ENSG00000000457   237.573      0.0295551 0.1292181 8.02723e-01 9.04035e-01
ENSG00000000460    58.035     -0.0507527 0.2117984 7.41002e-01 8.71680e-01
ENSG00000000971  5826.538      0.4057596 0.0892318 2.11372e-06 2.48262e-05
...                   ...            ...       ...         ...         ...
ENSG00000273448  14.02944      0.0178477  0.265789    0.904006    0.954503
ENSG00000273472  11.08483     -0.1020634  0.292642    0.463989    0.675274
ENSG00000273486  15.47814     -0.0468762  0.266725    0.747512    0.874323
ENSG00000273487   8.17605      0.2137027  0.373236    0.124802    0.292586
ENSG00000273488   8.59778      0.0225865  0.283616    0.860187    0.933030
```

# p 值校正

```{r}
resOrdered <- res[order(res$pvalue),]
summary(resOrdered)
```

```
out of 16596 with nonzero total read count
adjusted p-value < 0.1
LFC > 0 (up)       : 2618, 16%
LFC < 0 (down)     : 2298, 14%
outliers [1]       : 0, 0%
low counts [2]     : 0, 0%
(mean count < 5)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results
```

adjusted p value cutoff 原始設定為 `0.1`，若要修改則在 `results` 指定 `alpha` 參數

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

```
out of 16596 with nonzero total read count
adjusted p-value < 0.05
LFC > 0 (up)       : 2201, 13%
LFC < 0 (down)     : 1898, 11%
outliers [1]       : 0, 0%
low counts [2]     : 0, 0%
(mean count < 5)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results
```

# 繪圖

```{r}
png("./images/MA-plot.png", width = 800, height = 600, res = 120)
plotMA(res, ylim=c(-2,2))
dev.off()
```

![](./images/MA-plot.png)
