<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>fastqc 基本使用</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="fastqc_files/libs/clipboard/clipboard.min.js"></script>
<script src="fastqc_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="fastqc_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="fastqc_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="fastqc_files/libs/quarto-html/popper.min.js"></script>
<script src="fastqc_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="fastqc_files/libs/quarto-html/anchor.min.js"></script>
<link href="fastqc_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="fastqc_files/libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="fastqc_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="fastqc_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="fastqc_files/libs/bootstrap/bootstrap-91f297ac5b9b6094cb160b32f5ee6a0a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#連結" id="toc-連結" class="nav-link active" data-scroll-target="#連結"><span class="header-section-number">1</span> 連結</a></li>
  <li><a href="#fastqc-版本資訊" id="toc-fastqc-版本資訊" class="nav-link" data-scroll-target="#fastqc-版本資訊"><span class="header-section-number">2</span> fastqc 版本資訊</a></li>
  <li><a href="#執行-fastqc" id="toc-執行-fastqc" class="nav-link" data-scroll-target="#執行-fastqc"><span class="header-section-number">3</span> 執行 fastqc</a>
  <ul class="collapse">
  <li><a href="#pair-end-fastq-檔案" id="toc-pair-end-fastq-檔案" class="nav-link" data-scroll-target="#pair-end-fastq-檔案"><span class="header-section-number">3.1</span> pair-end FASTQ 檔案</a></li>
  </ul></li>
  <li><a href="#fastqc-指令參考" id="toc-fastqc-指令參考" class="nav-link" data-scroll-target="#fastqc-指令參考"><span class="header-section-number">4</span> fastqc 指令參考</a></li>
  <li><a href="#conda-環境資訊" id="toc-conda-環境資訊" class="nav-link" data-scroll-target="#conda-環境資訊"><span class="header-section-number">5</span> Conda 環境資訊</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">fastqc 基本使用</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="連結" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> 連結</h1>
<p><a href="https://github.com/benson1231/tools/tree/main/fastqc">create conda envs</a><br>
<a href="https://github.com/s-andrews/FastQC">fastqc official github</a></p>
</section>
<section id="fastqc-版本資訊" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> fastqc 版本資訊</h1>
<p>使用 <code>-v</code> 參數確認 fastqc 是否成功安裝，並顯示版本號。</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-v</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>FastQC v0.12.1</code></pre>
</div>
</div>
</section>
<section id="執行-fastqc" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> 執行 fastqc</h1>
<p>以下示範 fastqc 的基本用法：對單一或雙端 FASTQ 做品質檢查並輸出 HTML 與 ZIP 報告。</p>
<p>請確認：</p>
<ul>
<li>測試 FASTQ 位於 <code>../.test/</code></li>
</ul>
<p>輸出內容包含：</p>
<ul>
<li>HTML QC 報告（互動式）</li>
<li>ZIP 壓縮資料（含 QC 原始數據）</li>
</ul>
<section id="pair-end-fastq-檔案" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="pair-end-fastq-檔案"><span class="header-section-number">3.1</span> pair-end FASTQ 檔案</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>../.test/a.chr21_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>../.test/a.chr21_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>--outdir results/</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>application/gzip
Started analysis of a.chr21_R1.fastq.gz
application/gzip
Approx 15% complete for a.chr21_R1.fastq.gz
Approx 35% complete for a.chr21_R1.fastq.gz
Approx 55% complete for a.chr21_R1.fastq.gz
Approx 80% complete for a.chr21_R1.fastq.gz
Analysis complete for a.chr21_R1.fastq.gz
Started analysis of a.chr21_R2.fastq.gz
Approx 15% complete for a.chr21_R2.fastq.gz
Approx 35% complete for a.chr21_R2.fastq.gz
Approx 60% complete for a.chr21_R2.fastq.gz
Approx 80% complete for a.chr21_R2.fastq.gz
Analysis complete for a.chr21_R2.fastq.gz</code></pre>
</div>
</div>
</section>
</section>
<section id="fastqc-指令參考" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> fastqc 指令參考</h1>
<p>顯示 fastqc 提供的主要參數與使用說明，適合作為快速查詢指令用途。</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--help</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

    fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of a number of different modules, each one of 
    which will help to identify a different potential type of problem in your
    data.
    
    If no files to process are specified on the command line then the program
    will start as an interactive graphical application.  If files are provided
    on the command line then the program will run with no user interaction
    required.  In this mode it is suitable for inclusion into a standardised
    analysis pipeline.
    
    The options for the program as as follows:
    
    -h --help       Print this help file and exit
    
    -v --version    Print the version of the program and exit
    
    -o --outdir     Create all output files in the specified output directory.
                    Please note that this directory must exist as the program
                    will not create it.  If this option is not set then the 
                    output file for each sequence file is created in the same
                    directory as the sequence file which was processed.
                    
    --casava        Files come from raw casava output. Files in the same sample
                    group (differing only by the group number) will be analysed
                    as a set rather than individually. Sequences with the filter
                    flag set in the header will be excluded from the analysis.
                    Files must have the same names given to them by casava
                    (including being gzipped and ending with .gz) otherwise they
                    won't be grouped together correctly.
                    
    --nano          Files come from nanopore sequences and are in fast5 format. In
                    this mode you can pass in directories to process and the program
                    will take in all fast5 files within those directories and produce
                    a single output file from the sequences found in all files.                    
                    
    --nofilter      If running with --casava then don't remove read flagged by
                    casava as poor quality when performing the QC analysis.
                   
    --extract       If set then the zipped output file will be uncompressed in
                    the same directory after it has been created. If --delete is 
                    also specified then the zip file will be removed after the 
                    contents are unzipped. 
                    
    -j --java       Provides the full path to the java binary you want to use to
                    launch fastqc. If not supplied then java is assumed to be in
                    your path.
                   
    --noextract     Do not uncompress the output file after creating it.  You
                    should set this option if you do not wish to uncompress
                    the output when running in non-interactive mode.
                    
    --nogroup       Disable grouping of bases for reads &gt;50bp. All reports will
                    show data for every base in the read.  WARNING: Using this
                    option will cause fastqc to crash and burn if you use it on
                    really long reads, and your plots may end up a ridiculous size.
                    You have been warned!
                    
    --min_length    Sets an artificial lower limit on the length of the sequence
                    to be shown in the report.  As long as you set this to a value
                    greater or equal to your longest read length then this will be
                    the sequence length used to create your read groups.  This can
                    be useful for making directly comaparable statistics from 
                    datasets with somewhat variable read lengths.

    --dup_length    Sets a length to which the sequences will be truncated when 
                    defining them to be duplicates, affecting the duplication and
                    overrepresented sequences plot.  This can be useful if you have
                    long reads with higher levels of miscalls, or contamination with
                    adapter dimers containing UMI sequences.

                    
    -f --format     Bypasses the normal sequence file format detection and
                    forces the program to use the specified format.  Valid
                    formats are bam,sam,bam_mapped,sam_mapped and fastq
                    

    --memory        Sets the base amount of memory, in Megabytes, used to process 
                    each file.  Defaults to 512MB.  You may need to increase this if
                    you have a file with very long sequences in it.
                
    --svg           Save the graphs in the report in SVG format.

    -t --threads    Specifies the number of files which can be processed
                    simultaneously.  Each thread will be allocated 250MB of
                    memory so you shouldn't run more threads than your
                    available memory will cope with, and not more than
                    6 threads on a 32 bit machine
                  
    -c              Specifies a non-default file which contains the list of
    --contaminants  contaminants to screen overrepresented sequences against.
                    The file must contain sets of named contaminants in the
                    form name[tab]sequence.  Lines prefixed with a hash will
                    be ignored.

    -a              Specifies a non-default file which contains the list of
    --adapters      adapter sequences which will be explicity searched against
                    the library. The file must contain sets of named adapters
                    in the form name[tab]sequence.  Lines prefixed with a hash
                    will be ignored.
                    
    -l              Specifies a non-default file which contains a set of criteria
    --limits        which will be used to determine the warn/error limits for the
                    various modules.  This file can also be used to selectively 
                    remove some modules from the output all together.  The format
                    needs to mirror the default limits.txt file found in the
                    Configuration folder.
                    
   -k --kmers       Specifies the length of Kmer to look for in the Kmer content
                    module. Specified Kmer length must be between 2 and 10. Default
                    length is 7 if not specified.
                    
   -q --quiet       Suppress all progress messages on stdout and only report errors.
   
   -d --dir         Selects a directory to be used for temporary files written when
                    generating report images. Defaults to system temp directory if
                    not specified.
                    
BUGS

    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk
    or in www.bioinformatics.babraham.ac.uk/bugzilla/
                   
    </code></pre>
</div>
</div>
</section>
<section id="conda-環境資訊" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Conda 環境資訊</h1>
<p>完整記錄當前執行的 Conda 環境，有助於分析重現、版本追蹤與錯誤排查。</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"======================================="</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">" Conda Environment: </span><span class="va">$(</span><span class="ex">conda</span> info <span class="at">--envs</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">'*'</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{print $1}'</span><span class="va">)</span><span class="st">"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"======================================="</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> list</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>=======================================
 Conda Environment: #
fastqc-env
=======================================
# packages in environment at /home/benson/miniforge3/envs/fastqc-env:
#
# Name                       Version          Build                 Channel
_libgcc_mutex                0.1              conda_forge           conda-forge
_openmp_mutex                4.5              2_gnu                 conda-forge
_r-mutex                     1.0.1            anacondar_1           conda-forge
alsa-lib                     1.2.15.1         hb03c661_0            conda-forge
binutils_impl_linux-64       2.45             default_hfdba357_104  conda-forge
bwidget                      1.10.1           ha770c72_1            conda-forge
bzip2                        1.0.8            hda65f42_8            conda-forge
c-ares                       1.34.6           hb03c661_0            conda-forge
ca-certificates              2025.11.12       hbd8a1cb_0            conda-forge
cairo                        1.18.4           h3394656_0            conda-forge
curl                         8.17.0           h4e3cde8_1            conda-forge
dart-sass                    1.59.1           ha770c72_0            conda-forge
deno                         2.3.1            hbf66b88_0            conda-forge
deno-dom                     0.1.41           h4768de7_0            conda-forge
esbuild                      0.27.2           hfc2019e_0            conda-forge
fastqc                       0.12.1           hdfd78af_0            bioconda
font-ttf-dejavu-sans-mono    2.37             hab24e00_0            conda-forge
font-ttf-inconsolata         3.000            h77eed37_0            conda-forge
font-ttf-source-code-pro     2.038            h77eed37_0            conda-forge
font-ttf-ubuntu              0.83             h77eed37_3            conda-forge
fontconfig                   2.15.0           h7e30c49_1            conda-forge
fonts-conda-ecosystem        1                0                     conda-forge
fonts-conda-forge            1                hc364b38_1            conda-forge
freetype                     2.14.1           ha770c72_0            conda-forge
fribidi                      1.0.16           hb03c661_0            conda-forge
gcc_impl_linux-64            15.2.0           hc5723f1_16           conda-forge
gfortran_impl_linux-64       15.2.0           h281d09f_16           conda-forge
giflib                       5.2.2            hd590300_0            conda-forge
graphite2                    1.3.14           hecca717_2            conda-forge
gsl                          2.7              he838d99_0            conda-forge
gxx_impl_linux-64            15.2.0           hda75c37_16           conda-forge
harfbuzz                     12.2.0           h15599e2_0            conda-forge
icu                          75.1             he02047a_0            conda-forge
kernel-headers_linux-64      5.14.0           he073ed8_3            conda-forge
keyutils                     1.6.3            hb9d3cd8_0            conda-forge
krb5                         1.21.3           h659f571_0            conda-forge
lcms2                        2.17             h717163a_0            conda-forge
ld_impl_linux-64             2.45             default_hbd61a6d_104  conda-forge
lerc                         4.0.0            h0aef613_1            conda-forge
libblas                      3.11.0           5_h4a7cf45_openblas   conda-forge
libcblas                     3.11.0           5_h0358290_openblas   conda-forge
libcups                      2.3.3            hb8b1518_5            conda-forge
libcurl                      8.17.0           h4e3cde8_1            conda-forge
libdeflate                   1.25             h17f619e_0            conda-forge
libedit                      3.1.20250104     pl5321h7949ede_0      conda-forge
libev                        4.33             hd590300_2            conda-forge
libexpat                     2.7.3            hecca717_0            conda-forge
libffi                       3.5.2            h9ec8514_0            conda-forge
libfreetype                  2.14.1           ha770c72_0            conda-forge
libfreetype6                 2.14.1           h73754d4_0            conda-forge
libgcc                       15.2.0           he0feb66_16           conda-forge
libgcc-devel_linux-64        15.2.0           hcc6f6b0_116          conda-forge
libgcc-ng                    15.2.0           h69a702a_16           conda-forge
libgfortran                  15.2.0           h69a702a_16           conda-forge
libgfortran-ng               15.2.0           h69a702a_16           conda-forge
libgfortran5                 15.2.0           h68bc16d_16           conda-forge
libglib                      2.86.3           h6548e54_0            conda-forge
libgomp                      15.2.0           he0feb66_16           conda-forge
libiconv                     1.18             h3b78370_2            conda-forge
libjpeg-turbo                3.1.2            hb03c661_0            conda-forge
liblapack                    3.11.0           5_h47877c9_openblas   conda-forge
liblzma                      5.8.1            hb9d3cd8_2            conda-forge
libnghttp2                   1.67.0           had1ee68_0            conda-forge
libopenblas                  0.3.30           pthreads_h94d23a6_4   conda-forge
libpng                       1.6.53           h421ea60_0            conda-forge
libsanitizer                 15.2.0           h90f66d4_16           conda-forge
libssh2                      1.11.1           hcf80075_0            conda-forge
libstdcxx                    15.2.0           h934c35e_16           conda-forge
libstdcxx-devel_linux-64     15.2.0           hd446a21_116          conda-forge
libstdcxx-ng                 15.2.0           hdf11a46_16           conda-forge
libtiff                      4.7.1            h9d88235_1            conda-forge
libuuid                      2.41.3           h5347b49_0            conda-forge
libuv                        1.51.0           hb03c661_1            conda-forge
libwebp-base                 1.6.0            hd42ef1d_0            conda-forge
libxcb                       1.17.0           h8a09558_0            conda-forge
libxcrypt                    4.4.36           hd590300_1            conda-forge
libzlib                      1.3.1            hb9d3cd8_2            conda-forge
make                         4.4.1            hb9d3cd8_2            conda-forge
ncurses                      6.5              h2d0b736_3            conda-forge
nodejs                       22.21.1          h4a9c4b4_0            conda-forge
openjdk                      25.0.1           h5755bd7_0            conda-forge
openssl                      3.6.0            h26f9b46_0            conda-forge
pandoc                       3.6.3            ha770c72_0            conda-forge
pango                        1.56.4           hadf4263_0            conda-forge
pcre2                        10.47            haa7fec5_0            conda-forge
perl                         5.32.1           7_hd590300_perl5      conda-forge
pixman                       0.46.4           h54a6638_1            conda-forge
pthread-stubs                0.4              hb9d3cd8_1002         conda-forge
quarto                       1.8.26           hbf95b10_0            conda-forge
r-base                       4.3.3            hc28aa72_21           conda-forge
r-base64enc                  0.1_3            r43hb1dbf0f_1007      conda-forge
r-bslib                      0.9.0            r43hc72bb7e_0         conda-forge
r-cachem                     1.1.0            r43hb1dbf0f_1         conda-forge
r-cli                        3.6.5            r43h93ab643_0         conda-forge
r-digest                     0.6.37           r43h0d4f4ea_0         conda-forge
r-ellipsis                   0.3.2            r43hb1dbf0f_3         conda-forge
r-evaluate                   1.0.5            r43hc72bb7e_0         conda-forge
r-fastmap                    1.2.0            r43ha18555a_1         conda-forge
r-fontawesome                0.5.3            r43hc72bb7e_0         conda-forge
r-fs                         1.6.6            r43h93ab643_0         conda-forge
r-glue                       1.8.0            r43h2b5f3a1_0         conda-forge
r-here                       1.0.1            r43hc72bb7e_3         conda-forge
r-highr                      0.11             r43hc72bb7e_1         conda-forge
r-htmltools                  0.5.8.1          r43ha18555a_1         conda-forge
r-jquerylib                  0.1.4            r43hc72bb7e_3         conda-forge
r-jsonlite                   2.0.0            r43h2b5f3a1_0         conda-forge
r-knitr                      1.50             r43hc72bb7e_0         conda-forge
r-lattice                    0.22_7           r43h2b5f3a1_0         conda-forge
r-lifecycle                  1.0.4            r43hc72bb7e_1         conda-forge
r-matrix                     1.6_5            r43he966344_1         conda-forge
r-memoise                    2.0.1            r43hc72bb7e_3         conda-forge
r-mime                       0.13             r43h2b5f3a1_0         conda-forge
r-png                        0.1_8            r43h21f035c_2         conda-forge
r-r6                         2.6.1            r43hc72bb7e_0         conda-forge
r-rappdirs                   0.3.3            r43hb1dbf0f_3         conda-forge
r-rcpp                       1.1.0            r43h93ab643_0         conda-forge
r-rcpptoml                   0.2.3            r43h93ab643_0         conda-forge
r-reticulate                 1.43.0           r43h3697838_0         conda-forge
r-rlang                      1.1.6            r43h93ab643_0         conda-forge
r-rmarkdown                  2.29             r43hc72bb7e_0         conda-forge
r-rprojroot                  2.1.1            r43hc72bb7e_0         conda-forge
r-sass                       0.4.10           r43h93ab643_0         conda-forge
r-tinytex                    0.57             r43hc72bb7e_0         conda-forge
r-withr                      3.0.2            r43hc72bb7e_0         conda-forge
r-xfun                       0.53             r43h3697838_0         conda-forge
r-yaml                       2.3.10           r43hdb488b9_0         conda-forge
readline                     8.3              h853b02a_0            conda-forge
sed                          4.9              h6688a6e_0            conda-forge
sysroot_linux-64             2.34             h087de78_3            conda-forge
tk                           8.6.13           noxft_ha0e22de_103    conda-forge
tktable                      2.10             h8d826fa_7            conda-forge
typst                        0.13.0           h53e704d_0            conda-forge
tzdata                       2025c            h8577fbf_0            conda-forge
xorg-libice                  1.1.2            hb9d3cd8_0            conda-forge
xorg-libsm                   1.2.6            he73a12e_0            conda-forge
xorg-libx11                  1.8.12           h4f16b4b_0            conda-forge
xorg-libxau                  1.0.12           hb03c661_1            conda-forge
xorg-libxdmcp                1.1.5            hb03c661_1            conda-forge
xorg-libxext                 1.3.6            hb9d3cd8_0            conda-forge
xorg-libxfixes               6.0.2            hb03c661_0            conda-forge
xorg-libxi                   1.8.2            hb9d3cd8_0            conda-forge
xorg-libxrandr               1.5.4            hb9d3cd8_0            conda-forge
xorg-libxrender              0.9.12           hb9d3cd8_0            conda-forge
xorg-libxt                   1.3.1            hb9d3cd8_0            conda-forge
xorg-libxtst                 1.2.5            hb9d3cd8_3            conda-forge
zstd                         1.5.7            hb78ec9c_6            conda-forge</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>