---
title: "bcftools 基本使用"
date: today
date-format: "YYYY-MM-DD"
---

# bcftools 介紹

BCFtools 專為處理 VCF 和 BCF 資料而設計。它將輸入檔案「-」視為標準輸入 (stdin)，並將輸出傳送到標準輸出 (stdout)，方便 Unix pipes 使用。

::: {.callout-note}
# 基本指令

bcftools [--version|--version-only] [--help] [COMMAND] [OPTIONS]
:::

::: {.callout-tip}
# Refernece

Twelve years of SAMtools and BCFtools
Petr Danecek, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, Thomas Keane, Shane A McCarthy, Robert M Davies, Heng Li
GigaScience, Volume 10, Issue 2, February 2021, giab008, https://doi.org/10.1093/gigascience/giab008
:::

---

# bcftools --version (版本)

```bash
bcftools --version
```

```
bcftools 1.23
Using htslib 1.23
Copyright (C) 2025 Genome Research Ltd.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

---

# bcftools view ()


---

# bcftools filter ()


---

# bcftools merge ()

---

# bcftools concat ()

---

# bcftools sort ()

---

# bcftools index ()

---

# bcftools stats ()

---

# bcftools annotate ()



---

# bcftools --help

```bash
bcftools --help
```

<details><summary><strong>Click to expand</strong></summary>
```
bcftools 1.23
Using htslib 1.23
Copyright (C) 2025 Genome Research Ltd.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
(ngs-tools) benson@Medcomasia:/mnt/c/Users/benson_lee/git/tools$ bcftools --help

Program: bcftools (Tools for variant calling and manipulating VCFs and BCFs)
License: GNU GPLv3+, due to use of the GNU Scientific Library
Version: 1.23 (using htslib 1.23)

Usage:   bcftools [--version|--version-only] [--help] <command> <argument>

Commands:

 -- Indexing
    index        index VCF/BCF files

 -- VCF/BCF manipulation
    annotate     annotate and edit VCF/BCF files
    concat       concatenate VCF/BCF files from the same set of samples
    convert      convert VCF/BCF files to different formats and back
    head         view VCF/BCF file headers
    isec         intersections of VCF/BCF files
    merge        merge VCF/BCF files files from non-overlapping sample sets
    norm         left-align and normalize indels
    plugin       user-defined plugins
    query        transform VCF/BCF into user-defined formats
    reheader     modify VCF/BCF header, change sample names
    sort         sort VCF/BCF file
    view         VCF/BCF conversion, view, subset and filter VCF/BCF files

 -- VCF/BCF analysis
    call         SNP/indel calling
    consensus    create consensus sequence by applying VCF variants
    cnv          HMM CNV calling
    csq          call variation consequences
    filter       filter VCF/BCF files using fixed thresholds
    gtcheck      check sample concordance, detect sample swaps and contamination
    mpileup      multi-way pileup producing genotype likelihoods
    polysomy     detect number of chromosomal copies
    roh          identify runs of autozygosity (HMM)
    stats        produce VCF/BCF stats

 -- Plugins (collection of programs for calling, file manipulation & analysis)
    41 plugins available, run "bcftools plugin -lv" to see a complete list

 Most commands accept VCF, bgzipped VCF, and BCF with the file type detected
 automatically even when streaming from a pipe. Indexed VCF and BCF will work
 in all situations. Un-indexed VCF and BCF and streams will work in most but
 not all situations.
```
</details>