---
title: "maftools"
author: "benson1231"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    self_contained: yes
    css: styles.css
    highlight: pygments
vignette: >
  %\VignetteIndexEntry{01: Summarize, Analyze, and Visualize MAF Files}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# 基本介紹與參考資料

`maftools` 是一個用於分析並視覺化 MAF 檔案的套件。

以下指令是根據[官方文檔](https://bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/maftools.html)範例撰寫，並新增個人的筆記內容，所有相關著作權屬於原作者(如下):

------------------------------------------------------------------------

***Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP. 2018. Maftools: efficient and comprehensive analysis of somatic variants in cancer. [Genome Research. PMID: 30341162](https://doi.org/10.1101/gr.239244.118)***

------------------------------------------------------------------------

# 下載 maftools 套件

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

if (!requireNamespace("maftools", quietly = TRUE)) {
    BiocManager::install("maftools", update = FALSE, ask = FALSE)
}

# create `results/` folder
if (!dir.exists("results")) dir.create("results", recursive = TRUE)
```

# 讀取與總結 maf 檔

- 可以使用 [vcf2maf](https://github.com/mskcc/vcf2maf) 將 vcf 轉 maf

## 讀取 maf 檔

```{r}
library(maftools)

#path to TCGA LAML MAF file
laml.maf = system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools') 
#clinical information containing survival information and histology. This is optional
laml.clin = system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools') 

laml = read.maf(maf = laml.maf, clinicalData = laml.clin)
```

## 查看 MAF object

```{r}
#Typing laml shows basic summary of MAF file.
laml
```

## 匯出 maf summary 檔案

```{r}
#Shows sample summry.
getSampleSummary(laml)
#Shows gene summary.
getGeneSummary(laml)
#shows clinical data associated with samples
getClinicalData(laml)
#Shows all fields in MAF
getFields(laml)
#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'results/laml')
```


# 視覺化

## 繪製 maf summary plot

```{r}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
```

## 繪製 oncoplot plot

```{r}
#oncoplot for top ten mutated genes.
oncoplot(maf = laml, top = 10)
```

## 繪製 Transition and Transversions

```{r}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
#plot titv summary
plotTiTv(res = laml.titv)
```

## 繪製胺基酸變化的 Lollipop plots

```{r}
#lollipop plot for DNMT3A, which is one of the most frequent mutated gene in Leukemia.
lollipopPlot(
  maf = laml,
  gene = 'DNMT3A',
  AACol = 'Protein_Change',
  showMutationRate = TRUE,
  labelPos = 882
)
```

### 自訂資料作為輸入

```{r}
#example data
my_data = data.frame(pos = sample.int(912, 15, replace = TRUE), count = sample.int(30, 15, replace = TRUE))
head(my_data)

lollipopPlot(data = my_data, gene = "DNMT3A")

plotProtein(gene = "TP53", refSeqID = "NM_000546")
```

## 繪製 Rainfall plots


```{r}
brca <- system.file("extdata", "brca.maf.gz", package = "maftools")
brca = read.maf(maf = brca, verbose = FALSE)

rainfallPlot(maf = brca, detectChangePoints = TRUE, pointSize = 0.4)
```

## 與 TCGA cohorts 比較 mutation load

```{r}
laml.mutload = tcgaCompare(maf = laml, cohortName = 'Example-LAML', logscale = TRUE, capture_size = 50)
```


## 繪製 VAF plot

```{r}
plotVaf(maf = laml, vafCol = 'i_TumorVAF_WU')
```


# 處理 copy-number 資料

## 讀取 GISTIC 的輸出檔

```{r}
gistic_res_folder <- system.file("extdata", package = "maftools")
laml.gistic = readGistic(gisticDir = gistic_res_folder, isTCGA = TRUE)

#GISTIC object
laml.gistic
```

## 視覺化 GISTIC 結果

### 繪製 GISTIC Chrom Plot

```{r}
gisticChromPlot(gistic = laml.gistic, markBands = "all")

# Co-gisticChromPlot
coGisticChromPlot(gistic1 = laml.gistic, gistic2 = laml.gistic, g1Name = "AML-1", g2Name = "AML-2", type = 'Amp')
```


### 繪製 Bubble Plot

```{r}
gisticBubblePlot(gistic = laml.gistic)
```

### 繪製 Onco Plot

```{r}
gisticOncoPlot(gistic = laml.gistic, clinicalData = getClinicalData(x = laml), clinicalFeatures = 'FAB_classification', sortByAnnotation = TRUE, top = 10)
```

## CBS segments

### 總結 chromosomal arm level CN

```{r}
laml.seg <- system.file("extdata", "LAML_CBS_segments.tsv.gz", package = "maftools")
segSummarize_results = segSummarize(seg = laml.seg)
```

### 視覺化 CBS segments

```{r}
tcga.ab.009.seg <- system.file("extdata", "TCGA.AB.3009.hg19.seg.txt", package = "maftools")
plotCBSsegments(cbsFile = tcga.ab.009.seg)
```


# 綜合分析

## 體細胞相互作用

```{r}
#exclusive/co-occurance event analysis on top 10 mutated genes. 
somaticInteractions(maf = laml, top = 25, pvalue = c(0.05, 0.1))
```

## 使用 cluster 偵測 cancer driver genes

```{r}
laml.sig = oncodrive(maf = laml, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore')

head(laml.sig)

plotOncodrive(res = laml.sig, fdrCutOff = 0.1, useFraction = TRUE, labelSize = 0.5)
```

## 新增 pfam Domains

```{R}
laml.pfam = pfamDomains(maf = laml, AACol = 'Protein_Change', top = 10)

#Protein summary (Printing first 7 columns for display convenience)
laml.pfam$proteinSummary[,1:7, with = FALSE]

#Domain summary (Printing first 3 columns for display convenience)
laml.pfam$domainSummary[,1:3, with = FALSE]
```

## Survival 分析

### 單一基因

```{r}
#Survival analysis based on grouping of DNMT3A mutation status
mafSurvival(maf = laml, genes = 'DNMT3A', time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)
```

### 找出導致存活率低的基因集

```{r}
#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups
prog_geneset = survGroup(maf = laml, top = 20, geneSetSize = 2, time = "days_to_last_followup", Status = "Overall_Survival_Status", verbose = FALSE)

print(prog_geneset)

mafSurvGroup(maf = laml, geneSet = c("DNMT3A", "FLT3"), time = "days_to_last_followup", Status = "Overall_Survival_Status")
```

## 比較兩個 cohort

```{r}
#Primary APL MAF
primary.apl = system.file("extdata", "APL_primary.maf.gz", package = "maftools")
primary.apl = read.maf(maf = primary.apl)
#Relapse APL MAF
relapse.apl = system.file("extdata", "APL_relapse.maf.gz", package = "maftools")
relapse.apl = read.maf(maf = relapse.apl)

#Considering only genes which are mutated in at-least in 5 samples in one of the cohort to avoid bias due to genes mutated in single sample.
pt.vs.rt <- mafCompare(m1 = primary.apl, m2 = relapse.apl, m1Name = 'Primary', m2Name = 'Relapse', minMut = 5)
print(pt.vs.rt)
```

### 森林圖

```{r}
forestPlot(mafCompareRes = pt.vs.rt, pVal = 0.1)
```

### 並排繪製兩個腫瘤圖

```{r}
genes = c("PML", "RARA", "RUNX1", "ARID1B", "FLT3")
coOncoplot(m1 = primary.apl, m2 = relapse.apl, m1Name = 'PrimaryAPL', m2Name = 'RelapseAPL', genes = genes, removeNonMutated = TRUE)
```

### 並排繪製兩個長條圖

```{r}
coBarplot(m1 = primary.apl, m2 = relapse.apl, m1Name = "Primary", m2Name = "Relapse")
```

### 合併兩個lollipop Plot

```{r}
lollipopPlot2(m1 = primary.apl, m2 = relapse.apl, gene = "PML", AACol1 = "amino_acid_change", AACol2 = "amino_acid_change", m1_name = "Primary", m2_name = "Relapse")
```

## 臨床富集分析

```{r}
fab.ce = clinicalEnrichment(maf = laml, clinicalFeature = 'FAB_classification')

#Results are returned as a list. Significant associations p-value < 0.05
fab.ce$groupwise_comparision[p_value < 0.05]

plotEnrichmentResults(enrich_res = fab.ce, pVal = 0.05, geneFontSize = 0.5, annoFontSize = 0.6)
```

## 藥物-基因交互作用

```{r}
dgi = drugInteractions(maf = laml, fontSize = 0.75)

dnmt3a.dgi = drugInteractions(genes = "DNMT3A", drugs = TRUE)

#Printing selected columns.
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]
```

## 致癌 pathway

```{r}
pws = pathways(maf = laml, plotType = 'treemap')

plotPathways(maf = laml, pathlist = pws)
```

## 腫瘤異質性分析

### 需要 mclust 套件

```{r}
if (!requireNamespace("mclust", quietly = TRUE)) {
    BiocManager::install("mclust", update = FALSE, ask = FALSE)
}

#Heterogeneity in sample TCGA.AB.2972
library("mclust")
```

### 腫瘤樣本的異質性

```{r}
tcga.ab.2972.het = inferHeterogeneity(maf = laml, tsb = 'TCGA-AB-2972', vafCol = 'i_TumorVAF_WU')

print(tcga.ab.2972.het$clusterMeans)

#Visualizing results
plotClusters(clusters = tcga.ab.2972.het)
```

### 忽略拷貝數變異區域的變異

```{r}
seg = system.file('extdata', 'TCGA.AB.3009.hg19.seg.txt', package = 'maftools')
tcga.ab.3009.het = inferHeterogeneity(maf = laml, tsb = 'TCGA-AB-3009', segFile = seg, vafCol = 'i_TumorVAF_WU')

#Visualizing results. Highlighting those variants on copynumber altered variants.
plotClusters(clusters = tcga.ab.3009.het, genes = 'CN_altered', showCNvars = TRUE)
```

## 突變特徵

### 需要 BSgenome.Hsapiens.UCSC.hg19

```{r}
if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)) {
    BiocManager::install("BSgenome.Hsapiens.UCSC.hg19", update = FALSE, ask = FALSE)
}

library("BSgenome.Hsapiens.UCSC.hg19")

laml.tnm = trinucleotideMatrix(maf = laml, prefix = 'chr', add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg19")
```

### APOBEC富集樣品與未富集樣品之間的差異

```{r}
plotApobecDiff(tnm = laml.tnm, maf = laml, pVal = 0.2)
```

### 簽名分析

```{r}
packages <- c("NMF", "pheatmap")

for (pkg in packages) {
    if (!requireNamespace(pkg, quietly = TRUE)) {
        install.packages(pkg)
    }
}

library('NMF')
laml.sign = estimateSignatures(mat = laml.tnm, nTry = 6, pConstant=0.01)

plotCophenetic(res = laml.sign)

laml.sig = extractSignatures(mat = laml.tnm, n = 3, pConstant=0.01)

#Compate against original 30 signatures 
laml.og30.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "legacy")

#Compate against updated version3 60 signatures 
laml.v3.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "SBS")

library('pheatmap')
pheatmap::pheatmap(mat = laml.og30.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")

maftools::plotSignatures(nmfRes = laml.sig, title_size = 1.2, sig_db = "SBS")


# #  3D 長條圖
# library("barplot3d")
# #Visualize first signature
# sig1 = laml.sig$signatures[,1]
# barplot3d::legoplot3d(contextdata = sig1, labels = FALSE, scalexy = 0.01, sixcolors = "sanger", alpha = 0.5)
```

# 變體註釋






























